core/src/main/java/com/linecorp/armeria/common/DefaultPingHandler.java

```diff
@@ -0,0 +1,141 @@
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.http2.DefaultHttp2PingFrame;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2FrameWriter;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.util.concurrent.Future;
+
+/**
+ * Default implementation of {@link PingHandler}.
+ */
+@NotThreadSafe
+public class DefaultPingHandler implements PingHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultPingHandler.class);
+
+    private final Http2FrameWriter frameWriter;
+    private final Http2PingRequestResponsePair pair = new Http2PingRequestResponsePair();
+    private final Random random = new Random();
+
+    @Nullable
+    private Future<?> pingTimeoutFuture;
+    @Nullable
+    private Future<?> pingWriteFuture;
+
+    private boolean started;
+
+    /**
+     * Handles ping event. TODO detailed javadoc.
+     */
+    public DefaultPingHandler(final Http2FrameWriter frameWriter) {
+        this.frameWriter = frameWriter;
+    }
+
+    @Override
+    public void start(ChannelHandlerContext ctx) {
+        checkState(!started, "DefaultPingHandler already started.");
+
+        started = true;
+        schedulePingWrite(ctx);
+    }
+
+    private void schedulePingWrite(ChannelHandlerContext ctx) {
+        checkIfAlreadyStarted();
+
+        final Channel channel = ctx.channel();
+        final EventLoop eventLoop = channel.eventLoop();
+        final Http2PingFrame newPingFrame = getNewPingFrame();
+
+        // TODO move to ServerConfig.
+        final long pingDelay = 5000;
+        final long pingAckTimeout = 1000;
+
+        pingWriteFuture = eventLoop.schedule(
+                () -> frameWriter.writePing(ctx, newPingFrame.ack(), newPingFrame.content(), ctx.newPromise()),
+                pingDelay, TimeUnit.MILLISECONDS)
+                                   .addListener(f -> pair.setRequestFrame(newPingFrame))
+                                   .addListener(f -> {
+                                       if (f.isSuccess()) {
+                                           pingTimeoutFuture = eventLoop.schedule(() -> {
+                                               ctx.fireUserEventTriggered(
+                                                       new Http2PingTimeoutEvent()
+                                               );
+                                           }, pingAckTimeout, TimeUnit.SECONDS);
+                                       } else {
+                                           // TODO send WriteFailed event instead of Object
+                                           ctx.fireUserEventTriggered(new Object());
+                                       }
+                                   });
+    }
```

Comment:
But mostly this is the important piece. The flow is:

1. We schedule and ping 
2. When the PING write is success full we schedule an another runnable to notify for `Http2PingTimeoutEvent `
3. If we indeed get PING ACK back then cancel the future created in (2).

Any suggestions that you can recommend to make this not look like a callback hell?

Reply:
I've thought of another approach that may look more simpler. Idea is:

1. We scheduleWithFixedDelay a PING write periodically (`pingDelay`).
2. We send PING if it's the time
3. On next schedule, we see if we got ACK for the ping sent in (2).

But the drawback for this approach is it takes a maximum of `pingDelay` to discover that we did not get ACK for the PING back.

Reply:
Another idea is:

When we receive `IdleStateEvent` emitted by  `IdleTimeoutHandler`, then we send a PING to see if the connection is still alive. If we do not get an ACK back or write failed then we can close the connection or do any other necessary action.

Reply:
How about using `scheduleWithFixedDelay()`, counting the number of send PINGs and received ACKs, and then disconnect if `Math.abs(numPingAcks - numPings)` exceeds a certain value?

Reply:
> When we receive `IdleStateEvent` emitted by  `IdleTimeoutHandler`, then we send a PING to see if the connection is still alive. If we do not get an ACK back or write failed then we can close the connection or do any other necessary action.

This also sounds good to me. How about combining `IdleTimeoutHandler` and PING/ACK counters?